src/python/pants/backend/experimental/python/lockfile_metadata.py

```diff
@@ -0,0 +1,77 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import hashlib
+import json
+from dataclasses import dataclass
+
+from pants.backend.python.util_rules.interpreter_constraints import InterpreterConstraints
+from pants.util.ordered_set import FrozenOrderedSet
+
+BEGIN_LOCKFILE_HEADER = b"# --- BEGIN PANTS LOCKFILE METADATA: DO NOT EDIT OR REMOVE ---"
+END_LOCKFILE_HEADER = b"# --- END PANTS LOCKFILE METADATA ---"
+
+
+@dataclass
+class LockfileMetadata:
+    invalidation_digest: str | None
+
+
+def invalidation_digest(
+    requirements: FrozenOrderedSet[str], interpreter_constraints: InterpreterConstraints
+) -> str:
+    """Returns an invalidation digest for the given requirements and interpreter constraints."""
+    m = hashlib.sha256()
+    pres = {
+        "requirements": [str(i) for i in requirements],
```

Comment:
They're already strings

```suggestion
        "requirements": list(requirements),
```